File: app/src/main/java/pl/szczodrzynski/edziennik/data/db/dao/FeedbackMessageDao.java
/*
 * Copyright (c) Kacper Ziubryniewicz 2020-1-6
 */

package pl.szczodrzynski.edziennik.data.db.dao;

import java.util.List;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;

import pl.szczodrzynski.edziennik.data.db.entity.FeedbackMessage;
import pl.szczodrzynski.edziennik.data.db.full.FeedbackMessageWithCount;

@Dao
public interface FeedbackMessageDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void add(FeedbackMessage feedbackMessage);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void addAll(List<FeedbackMessage> feedbackMessageList);

    @Query("DELETE FROM feedbackMessages")
    void clear();

    @Query("SELECT * FROM feedbackMessages")
    List<FeedbackMessage> getAllNow();

    @Query("SELECT * FROM feedbackMessages WHERE fromUser = :fromUser")
    List<FeedbackMessage> getAllByUserNow(String fromUser);

    @Query("SELECT * FROM feedbackMessages")
    LiveData<List<FeedbackMessage>> getAll();

    @Query("SELECT *, COUNT(*) AS messageCount FROM feedbackMessages GROUP BY fromUser ORDER BY sentTime DESC")
    List<FeedbackMessageWithCount> getAllWithCountNow();
}

File: app/src/main/java/pl/szczodrzynski/edziennik/data/db/dao/FeedbackMessageDao.kt
/*
 * Copyright (c) Kacper Ziubryniewicz 2020-1-6
 */
package pl.szczodrzynski.edziennik.data.db.dao

import androidx.lifecycle.LiveData
import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import pl.szczodrzynski.edziennik.data.db.entity.FeedbackMessage

@Dao
interface FeedbackMessageDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun add(feedbackMessage: FeedbackMessage)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun addAll(feedbackMessageList: List<FeedbackMessage>)

    @Query("DELETE FROM feedbackMessages")
    fun clear()

    @get:Query("SELECT * FROM feedbackMessages ORDER BY sentTime DESC LIMIT 50")
    val allNow: List<FeedbackMessage>

    @Query("SELECT * FROM feedbackMessages WHERE deviceId = :deviceId ORDER BY sentTime DESC LIMIT 50")
    fun getByDeviceIdNow(deviceId: String): List<FeedbackMessage>

    @get:Query("SELECT * FROM feedbackMessages")
    val all: LiveData<List<FeedbackMessage>>

    @get:Query("SELECT *, COUNT(*) AS count FROM feedbackMessages WHERE received = 1 AND devId IS NULL AND deviceId != 'szkolny.eu' GROUP BY deviceId ORDER BY sentTime DESC")
    val allWithCountNow: List<FeedbackMessage.WithCount>
}
